### PR TITLE
Lesson03

### PR DESCRIPTION
**Pull request by Polya404**

### CODE REVIEW

File: lesson_03/src/main/java/org/application/App.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package org.application;
```
Comment (kartius) [inline lesson_03/src/main/java/org/application/App.java:1]: Please, add to your PR only home task

File: homework_03/src/main/java/org/application/ValueCalculator.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    private int numOfThreads = 1;
+
+    public void setNumOfThreads(int numOfThreads) {
+        this.numOfThreads = numOfThreads;
+    }
+
+    public void operationsWithMultipleThreads() throws InterruptedException {
+        long start = System.currentTimeMillis();
+        Arrays.fill(array, 1f);
+        CountDownLatch latch = new CountDownLatch(numOfThreads);
```
Comment (kartius) [inline homework_03/src/main/java/org/application/ValueCalculator.java]: Why did you use CountDownLatch here? It's no necessary

File: homework_03/src/main/java/org/application/ValueCalculator.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    private void shuffleArray() {
+        List<Float> list = List.of(array);
+        Collections.shuffle(List.of(list));
+        for (int i = 0; i < list.size(); i++) {
+            array[i] = list.get(i);
+        }
+    }
+
+    private void findMaxValue() {
+        Float max = Collections.max(List.of(array));
```
Comment (kartius) [inline homework_03/src/main/java/org/application/ValueCalculator.java]: The recognizing of max value isn't in parallel calculation
You need to add extra logic to split array into several parts and each thread should find the max value in this part and then compare all values and choose the maximum

File: homework_03/src/main/java/org/application/App.java [hunk 4]
```diff
@@ -1,9 +1,11 @@
 package org.application;
 public class App 
 {
-    public static void main( String[] args ) throws InterruptedException {
+    public static void main( String[] args ) {
         ValueCalculator calculator = new ValueCalculator();
```
Comment (Polya404) [inline homework_03/src/main/java/org/application/App.java:5]: Perhaps I did something wrong, I don’t quite understand why 10 threads work slower than 1 and why for an array with 10 elements a large number of threads work faster compared to one thread than for an array with 1000000 elements

Comment (kartius) [reply homework_03/src/main/java/org/application/App.java:5]: You make this logic 
```
          int startIndex = i * segmentSize;
            int endIndex = (i == numOfThreads - 1) ? ARRAY_SIZE : (i + 1) * segmentSize;
            Float[] subArray = Arrays.copyOfRange(array, startIndex, endIndex);
```
not in thread internally
If you have 1 thread amount the code above will be called once
If you have 10 threads amount the code will be called 10 times

Comment (kartius) [reply homework_03/src/main/java/org/application/App.java:5]: Please, move the logic out of foreach and re-check perfomance

**Review by kartius — APPROVED** [review]

**Pull request merged**